File: 0528-swapping-nodes-in-a-linked-list/0528-swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        i = 1
        fast = slow = head
        prev_fast = None
        prev_slow = None
        while i < k:
            prev_fast = fast
            fast = fast.next
            i += 1
        begin_node = fast
        while fast and fast.next:
            prev_slow = slow
            slow = slow.next
            fast = fast.next
        end_node = slow
        if begin_node == end_node:
            return head
        if prev_slow:
            prev_slow.next = begin_node
        else:
            head = fast

        if prev_fast:
            prev_fast.next = end_node
        else:
            head = slow

        fast_next_node = begin_node.next
        slow_next_node = end_node.next
        begin_node.next = slow_next_node
        end_node.next = fast_next_node
        return head

        

File: 0528-swapping-nodes-in-a-linked-list/0528-swapping-nodes-in-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) 
    {
        if(head==null || head.next==null) return head;
         ListNode start = head;
        // Move start to the k-th node from the beginning
        for (int i = 1; i < k; i++) {
            start = start.next;
        }

        ListNode end = head;
        ListNode dummy = start;

        // Move dummy to the end, moving end simultaneously
        while (dummy.next != null) {
            dummy = dummy.next;
            end = end.next;
        }

        // Swap values of start and end nodes
        int temp = start.val;
        start.val = end.val;
        end.val = temp;

        return head;
        
    }
}